File: problem21/p21.py
import math


def sum_of_divisors(number):
    total = 1
    for i in range(2, int(math.sqrt(number)) + 1):
        if number % i == 0:
            total += i + (number//i)
    return total


def amicable_numbers(limit):
    result = 0
    number_list = [sum_of_divisors(i) for i in range(limit + 1)]
    for i in range(1, limit):
        if 0 < number_list[i] < limit:
            if i == number_list[number_list[i]]:
                if i != number_list[i]:
                    result += i + number_list[i]
                    number_list[number_list[i]] = 0
    return result


print(amicable_numbers(10000))

File: problem21/p21.java
public class p21 {
	
	public static int sum_of_divisors(int number){
	    int i, sum = 1;
	    for(i=2; i<=Math.sqrt(number); i++){
	        if(number % i == 0){
	            sum += i + (number/i);
	        }
	    }
	    return sum;
	}

	public static int amicable_numbers(int limit){
	    int i, sum = 0;
	    int[] numberList = new int[limit];
	    for(i=1; i<limit; i++){
	        numberList[i] = sum_of_divisors(i);
	    }
	    for(i=1; i<limit; i++){
	        if(numberList[i] > 0 && numberList[i] < limit){
	            if(i == numberList[numberList[i]]){
	                if(i != numberList[i]){
	                    sum += i + numberList[i];
	                    numberList[numberList[i]] = 0;
	                }
	            }
	        }
	    }

	    return sum;
	}

	public static void main(String args[]) {
		System.out.println(amicable_numbers(10000));
	}
	
}
